hooks/basic/create_version.py

```diff
@@ -231,24 +249,102 @@ def publish(self, task_settings, item):
         # set review_submission app's env/context based on item (ingest)
         self.__review_submission_app.change_context(item.context)
 
-        sg_version = self.__review_submission_app.render_and_submit_path(
-            item.properties.path,
-            fields,
-            first_frame,
-            last_frame,
-            sg_publish_data,
-            item.context.task,
-            item.description,
-            item.get_thumbnail_as_path(),
-            self._progress_cb,
-            colorspace
-        )
+        exception = None
+        sg_version = None
+        try:
+
+            # let's check if the pre-processed paths also contain rendered the movie
+            if pre_processed_paths:
+                movie_path = self._get_movie_path(task_settings, item)
```

Comment:
Why not:
```
if movie_path not in pre_processed_paths:
   render()

submit_version(movie_path)
```

Reply:
We can't have multiple configurations for review submit. eg. if the color processing hook ran review submit in the previous step, the pre processed paths property should contain the movie path too, otherwise there is no other way for us to reconfigure the review submit for different hooks.

Reply:
Umm...still don't follow why my logic wouldn't also satisfy that.

Reply:
There can only be one configuration of review submit per role, if it didn't generate the movie in the first run, it won't in the second run either because the nuke hook isn't configured to write out a movie. The above scenario should never happen but I have just added that check as a safety-net.

Reply:
Ok...I ran through this again and now have a better understanding of the problem. Ultimately, though, its really sitting wrong with me that one plugin is impinging on the domain of another. We really should enforce atomic operations as much as possible.

So...while you're right that we cannot configure `tk-multi-reviewsubmission` on a per-task basis, we can instantiate multiple instances of `tk-multi-reviewsubmission` for a particular engine. What I would suggest doing then, is creating an override for the `tk-shell` engine in the sgtk_config_environments_ingest.yaml file, and adding a second `tk-multi-reviewsubmission` instance (maybe called `tk-multi-rendercolornodes`) that then gets referenced by the ColorProcessFiles plugin and is configured to run a different nuke script that just renders out the color nodes. This would allow us to basically get rid of all the `pre_processed_paths` and `extra_write_node_mapping` wonkiness and would maintain atomicity.

Reply:
Ingest role already overrides the settings of `tk-multi-reviewsubmission` enabling a different instance still won't give us any advantage over the way we are doing it right now.

In the current method the advantage is we are only subprocessing nuke once in the color processing hook, and the create version just uses the movie that came as a result of the previous hook and creates a version out of it, instead of having 2 different nuke subprocesses one which shoots a movie(in create version) and other that processes frames (color processing hook). Thoughts?

Reply:
The additional Nuke subprocess argument is not enough to sway me. If its really a concern, then we should look at methods to parallelize non-dependent Task execution. The additional instance provides the advantage of having two completely unique instances each of which can be tailored specifically to the task at hand. For example, besides changing the Nuke script with different outputs/products, perhaps you want to use different codecs or different resolutions for Ingest from what you use for CreateVersion. Plus, again, there is the added advantage of atomicity, which I would argue reduces more tech debt than my solution would otherwise create.

Reply:
We can think of  parallelizing those tasks but it's not necessary, since we will need the color processed files before the movie creation. Which was already being handled in the nuke hook used by the color processing plugin, previously. i.e nuke hook was running the color conversion step first and then movie creation.

I have added the new app instance with separate settings from reviewsubmission app, instance is called `tk-multi-colorprocessfiles`.
Using a different resolution or codec for the movie creation is a valid use-case for it. But we were achieving that by changing those in the role based config files as well.
The color processing nuke script is given to us by the comp supe's since those are different for each show, we will now have to sit with them to help them create two nuke scripts and their corresponding nuke hooks, one which processes ACEScg files to movies, which will be used by `tk-multi-reviewsubmission` . Second nuke script which converts the client images to ACEScg which will be used by `tk-multi-colorprocessfiles` instance of review submit.

Reply:
Yeah, that's kind of expected, and is part of the process for Supe's here as well.